dynamiq/memory/backends/postgresql.py

```diff
@@ -0,0 +1,305 @@
+            sql += " LIMIT %s"
+            params.append(limit)
+
+        rows = self._execute_sql(sql, params, fetch="all")
+        messages = [self._row_to_message(row) for row in rows]
+        logger.debug(f"PostgreSQL Memory ({self.table_name}): Retrieved {len(messages)} messages.")
+        return messages
+
+    def _build_where_clause(self, query: str | None, filters: dict | None) -> tuple[str, list]:
+        """Builds the WHERE clause and parameters for search."""
+        where_clauses = []
+        params = []
+
+        if filters:
+            for key, value in filters.items():
+                where_clauses.append(f'"{self.metadata_col}"->>%s = %s')
+                params.extend([key, str(value)])  # Compare as text
+        if query:
+            where_clauses.append(f'"{self.content_col}" ILIKE %s')
+            params.append(f"%{query}%")
+        if not where_clauses:
+            return "", []
+
+        where_clause_str = "WHERE " + " AND ".join(where_clauses)
+        return where_clause_str, params
+
+    def search(
+        self, query: str | None = None, filters: dict[str, Any] | None = None, limit: int | None = None
+    ) -> list[Message]:
+        """Searches messages using ILIKE for query and JSONB operators for filters."""
+
+        where_clause, params = self._build_where_clause(query, filters)
+
+        sql = self._SELECT_SEARCH_SQL.format(
+            table_name=self.table_name,
+            message_id_col=self.message_id_col,
+            role_col=self.role_col,
+            content_col=self.content_col,
+            metadata_col=self.metadata_col,
+            timestamp_col=self.timestamp_col,
+        )
+
+        if where_clause:
+            sql += f" {where_clause}"
+
+        sql += f' ORDER BY "{self.timestamp_col}" DESC'
+
+        if limit is not None and limit > 0:
+            sql += " LIMIT %s"
+            params.append(limit)
+
+        rows = self._execute_sql(sql, params, fetch="all")
+        messages = [self._row_to_message(row) for row in rows]
+
+        logger.debug(
+            f"PostgreSQL Memory ({self.table_name}): Found {len(messages)} search results "
+            f"(Query: {'Yes' if query else 'No'}, Filters: {'Yes' if filters else 'No'}, Limit: {limit})"
+        )
+        return messages
+
+    def is_empty(self) -> bool:
+        """Checks if the PostgreSQL table is empty."""
+        sql = self._IS_EMPTY_SQL.format(table_name=self.table_name)
+        result = self._execute_sql(sql, fetch="one")
+        return not (result and result.get("exists", False))
+
+    def clear(self) -> None:
+        """Clears the PostgreSQL table using TRUNCATE."""
+        logger.warning(f"Clearing all messages from PostgreSQL table '{self.table_name}' using TRUNCATE.")
+        sql = self._CLEAR_SQL.format(table_name=self.table_name)
+        self._execute_sql(sql)
+        logger.info(f"PostgreSQL Memory ({self.table_name}): Cleared table.")
+
+    def __del__(self):
```

Comment:
- Provide a close() method in your class and call it explicitly when you are done with the connection
- Generally, __del__ is not recommended for such cases. Context managers is the most Pythonic way to handle resources that need to be closed. But I'm not sure if this fit your use case